### PR TITLE
TC: Add missing test cases of Libc misc

### PR DESCRIPTION
**Pull request by lokeshbv**
Adds tc for lib_dumpbuffer, sendfile and match api's.

NOTE: As lib_dumpbuffer return void, the success or failure result should be
checked manually by looking into tc output.
It has 2 cases,
1. dumps "53414d53554e472d54697a656e5254 SAMSUNG-TizneRT"
2. dumps "0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a ................................" (i.e, 256 "0a")

Signed-off-by: Lokesh B V <lokesh.bv@partner.samsung.com>

### CODE REVIEW

File: apps/examples/testcase/le_tc/kernel/tc_libc_misc.c [hunk 1]
```diff
@@ -347,6 +348,75 @@ static void tc_libc_misc_vdbg(void)
...
+{
+	const char *msg = "tc_libc_misc_dumpbuffer";
+	unsigned char buffer[] = { 'S', 'A', 'M', 'S', 'U', 'N', 'G', '-', 'T', 'i', 'z', 'e', 'n', 'R', 'T' };
+	unsigned char *buf = NULL;
+	int ret_chk;
+	int idx;
+
+	/* As lib_dumpbuffer returns void, there in no way to check the success or failure case
+	* we need to manually check the dump output
+	* Case-1: output: SRIB-TizenRT
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/kernel/tc_libc_misc.c]: please check Case-1 output

Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/kernel/tc_libc_misc.c]: and when I tested, the printed result is below

>tc_libc_misc_dumpbuffer (2033fc0):
0000: 53414d53554e472d54697a656e5254                                    SAMSUNG-TizenRT
tc_libc_misc_dumpbuffer (20316c0):
0000: 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a ................ ................
0020: 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a ................ ................
0040: 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a ................ ................
0060: 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a ................ ................
0080: 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a ................ ................
00a0: 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a ................ ................
00c0: 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a ................ ................
00e0: 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a ................ ................

File: apps/examples/testcase/le_tc/kernel/tc_libc_misc.c [hunk 2]
```diff
@@ -347,6 +348,75 @@ static void tc_libc_misc_vdbg(void)
...
+	* Case-2: output: 0a0a0a0a0a0a0a...........
+	*/
+
+	/* Case-1 */
+	lib_dumpbuffer(msg, buffer, sizeof(buffer));
+
+	/* Case-2 */
+	buf = malloc(BUFF_SIZE);
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/kernel/tc_libc_misc.c]: type casting to unsigned char * is needed for build warnings

Comment (lokeshbv) [inline apps/examples/testcase/le_tc/kernel/tc_libc_misc.c]: done

File: apps/examples/testcase/le_tc/kernel/tc_libc_misc.c (hunk 1 above)
Comment (lokeshbv) [inline apps/examples/testcase/le_tc/kernel/tc_libc_misc.c]: done

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 3]
```diff
@@ -1014,17 +1017,73 @@ static void fs_vfs_sendfile_tc(void)
 	TC_ASSERT_EQ_CLEANUP("stat", ret, OK, close(fd1));
 
 	size = st.st_size;
+    /* case-1: offset = 0 */
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/filesystem/fs_main.c]: please change the space to tab

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 4]
```diff
@@ -1014,17 +1017,73 @@ static void fs_vfs_sendfile_tc(void)
...
 	fd2 = open(dest_file, O_WRONLY | O_CREAT);
 	TC_ASSERT_GEQ_CLEANUP("open", fd2, 0, close(fd1));
 
 	ret = sendfile(fd2, fd1, 0, size);
-	close(fd1);
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/filesystem/fs_main.c]: why did you remove close(fd1)?
I think it is needed

Comment (lokeshbv) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: Its not removed, its there down at line 1086 , please check it.

File: apps/examples/testcase/le_tc/filesystem/fs_main.c (hunk 3 above)
Comment (lokeshbv) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: done

File: apps/examples/testcase/le_tc/filesystem/fs_main.c (hunk 4 above)
Comment (sunghan-chang) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: Before line 1086, there are many TC_ASSERT_XX macro. If some failure is there, this tc will be returned without "close(fd1)". This is a leak.

Comment (lokeshbv) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: yes, I missed to check, I will update.

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 5]
```diff
@@ -1014,17 +1017,73 @@ static void fs_vfs_sendfile_tc(void)
...
+	/* case-4: invalid input fd, returns ERROR */
+	fd2 = open(dest_file, O_WRONLY | O_CREAT);
+	TC_ASSERT_GEQ_CLEANUP("open", fd2, 0, close(fd1));
+
+	ret = sendfile(fd2, INV_FD, NULL, size);
+	close(fd2);
+	TC_ASSERT_EQ_CLEANUP("sendfile", ret, ERROR, close(fd1));
+
+	/* case-5: offset = 0, invalid output fd, returns ERROR */
+	fd2 = open(dest_file, O_WRONLY | O_CREAT);
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/filesystem/fs_main.c]: fd2 is not used in this case. Do we need to open fd2?

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 6]
```diff
@@ -1014,17 +1017,73 @@ static void fs_vfs_sendfile_tc(void)
...
+	ret = sendfile(INV_FD, fd1, &offset, size);
+	close(fd2);
+	TC_ASSERT_EQ_CLEANUP("sendfile", ret, ERROR, close(fd1));
+
+	/* case-6: current offset of input file is EOF, returns ERROR */
+	fd2 = open(dest_file, O_WRONLY | O_CREAT);
+	TC_ASSERT_GEQ_CLEANUP("open", fd2, 0, close(fd1));
+
+	ret = lseek(fd1, 0, SEEK_END);
+	TC_ASSERT_EQ_CLEANUP("lseek", ret, size, close(fd1));
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/filesystem/fs_main.c]: fd2 is already opened so that we should add "close(fd2)" also.
```
TC_ASSERT_EQ_CLEANUP("lseek", ret, size, close(fd1));
=>
TC_ASSERT_EQ_CLEANUP("lseek", ret, size, close(fd1);\close(fd2));
```

File: apps/examples/testcase/le_tc/filesystem/fs_main.c (hunk 5 above)
Comment (lokeshbv) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: No, Fixed it

File: apps/examples/testcase/le_tc/filesystem/fs_main.c (hunk 6 above)
Comment (lokeshbv) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: Fixed it

File: apps/examples/testcase/le_tc/kernel/tc_libc_misc.c [hunk 7]
```diff
@@ -347,6 +348,76 @@ static void tc_libc_misc_vdbg(void)
...
+	 */
+
+	/* Case-1 */
+	lib_dumpbuffer(msg, buffer, sizeof(buffer));
+
+	/* Case-2 */
+	buf = (unsigned char *)malloc(BUFF_SIZE);
+	if (!buf) {
+		ret_chk = dbg("malloc failed");
+		TC_ASSERT_EQ("dbg", ret_chk, strlen("malloc failed"));
```
Comment (jeongarmy) [inline apps/examples/testcase/le_tc/kernel/tc_libc_misc.c]: I think that it seems to be unnecessary.
Why do you check whether malloc is failed or not using dbg in this tc?
TC_ASSERT("xxx", buf) is enough

**Comment by sunghan-chang** [comment]
@lokeshbv Could you check @jeongarmy 's comments?

**Comment by lokeshbv** [comment]
@sunghan-chang @jeongarmy , sorry for the delay, I was on sick leave.
Yes, no need to do malloc validation again, if we use TC_ASSERT, I have updated the changes. Please review the same.

File: apps/examples/testcase/le_tc/kernel/tc_libc_misc.c [hunk 8]
```diff
@@ -347,6 +349,72 @@ static void tc_libc_misc_vdbg(void)
...
+	 * Case-2: output: 0a0a0a0a0a0a0a0a0a0a0a0a0a0a0a ............... (i.e, 256 "0a")
+	 */
+
+	/* Case-1 */
+	lib_dumpbuffer(msg, buffer, sizeof(buffer));
+
+	/* Case-2 */
+	buf = (unsigned char *)malloc(BUFF_SIZE);
+	TC_ASSERT("malloc", buf);
+
```
Comment (jeongarmy) [inline apps/examples/testcase/le_tc/kernel/tc_libc_misc.c]: We will remove TC_ASSERT. please use TC_ASSERT_EQ/NEQ instead of TC_ASSERT for readability.
For example, TC_ASSERT_EQ("malloc", buf, NULL);

File: apps/examples/testcase/le_tc/kernel/tc_libc_misc.c [hunk 9]
```diff
@@ -347,6 +349,72 @@ static void tc_libc_misc_vdbg(void)
...
+ * @API's covered       :lib_match
+ * @Preconditions       :None
+ * @Postconditions      :None
+ * @Return              :void
+ */
+static void tc_libc_misc_match(void)
+{
+	const char *pattern = "?|";
+	const char *string = "S";
+	int ret;
```
Comment (jeongarmy) [inline apps/examples/testcase/le_tc/kernel/tc_libc_misc.c]: Two words, "?l" and "S" are assigned to const region becasue you used const keyword.
so these will not disappear after this function is finished.
And we can't access these values after we assign new value to pattern and string variable in line 410

File: apps/examples/testcase/le_tc/kernel/tc_libc_misc.c (hunk 8 above)
Comment (lokeshbv) [reply apps/examples/testcase/le_tc/kernel/tc_libc_misc.c]: Ok

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 10]
```diff
@@ -1016,17 +1019,69 @@ static void fs_vfs_sendfile_tc(void)
...
+	TC_ASSERT_GEQ_CLEANUP("open", fd2, 0, close(fd1));
+
+	ret = lseek(fd1, 0, SEEK_END);
+	TC_ASSERT_EQ_CLEANUP("lseek", ret, size,  close(fd1);close(fd2));
+
+	ret = sendfile(fd2, fd1, NULL, size);
+	close(fd2);
+	TC_ASSERT_EQ_CLEANUP("sendfile", ret, 0, close(fd1));
+
+	ret = stat(dest_file, &st);
```
Comment (jeongarmy) [inline apps/examples/testcase/le_tc/filesystem/fs_main.c]: If you close fd1 here, you can reduce unnessasary duplicated codes for cleanup.
please move close(fd1) up.

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 11]
```diff
@@ -1016,17 +1019,69 @@ static void fs_vfs_sendfile_tc(void)
...
+	TC_ASSERT_GEQ_CLEANUP("open", fd2, 0, close(fd1));
+
+	offset = 1;
+	ret = sendfile(fd2, fd1, &offset, size - 1);
+	close(fd2);
+	TC_ASSERT_EQ_CLEANUP("sendfile", ret, size - 1, close(fd1));
+
+	ret = stat(dest_file, &st);
+	TC_ASSERT_EQ_CLEANUP("stat", ret, OK, close(fd1));
+	TC_ASSERT_EQ_CLEANUP("stat", size - 1, st.st_size, close(fd1));
```
Comment (jeongarmy) [inline apps/examples/testcase/le_tc/filesystem/fs_main.c]: Please exchange size-1 and st.st_size.

File: apps/examples/testcase/le_tc/kernel/tc_libc_misc.c (hunk 9 above)
Comment (lokeshbv) [reply apps/examples/testcase/le_tc/kernel/tc_libc_misc.c]: Yes, I agree with your point that 2 string literals will be there in read-only data section, even after function is finished.
1. Why do you want to access used string literals?
2. If your concern is about using string literals, I can provided better approach to that using char array.

Please elaborate your concern/question regarding this comment.

File: apps/examples/testcase/le_tc/filesystem/fs_main.c (hunk 11 above)
Comment (lokeshbv) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: ok!

File: apps/examples/testcase/le_tc/filesystem/fs_main.c (hunk 10 above)
Comment (lokeshbv) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: correct, will update!

File: apps/examples/testcase/le_tc/kernel/tc_libc_misc.c (hunk 9 above)
Comment (jeongarmy) [reply apps/examples/testcase/le_tc/kernel/tc_libc_misc.c]: Sorry, I misundersood for a moment. you don't need to fix it

File: apps/examples/testcase/le_tc/filesystem/fs_main.c [hunk 12]
```diff
@@ -1018,17 +1021,69 @@ static void fs_vfs_sendfile_tc(void)
...
 
 	ret = sendfile(fd2, fd1, 0, size);
-	close(fd1);
 	close(fd2);
-	TC_ASSERT_EQ("sendfile", ret, size);
+	TC_ASSERT_EQ_CLEANUP("sendfile", ret, size, close(fd1));
 
 	ret = stat(dest_file, &st);
+	TC_ASSERT_EQ_CLEANUP("stat", ret, OK, close(fd1));
+	TC_ASSERT_EQ_CLEANUP("stat", size, st.st_size, close(fd1));
```
Comment (jeongarmy) [inline apps/examples/testcase/le_tc/filesystem/fs_main.c]: please exchange size and st.st_size

Comment (lokeshbv) [reply apps/examples/testcase/le_tc/filesystem/fs_main.c]: done

File: apps/examples/testcase/le_tc/kernel/tc_libc_misc.c [hunk 13]
```diff
@@ -347,6 +349,72 @@ static void tc_libc_misc_vdbg(void)
...
+{
+	const char *msg = "tc_libc_misc_dumpbuffer";
+	unsigned char buffer[] = { 'S', 'A', 'M', 'S', 'U', 'N', 'G', '-', 'T', 'i', 'z', 'e', 'n', 'R', 'T' };
+	unsigned char *buf = NULL;
+	int idx;
+
+	/**
+	 * As lib_dumpbuffer returns void, there in no way to check the success or failure case
+	 * we need to manually check the dump output
+	 * Case-1: output: 53414d53554e472d54697a656e5254 SAMSUNG-TizenRT
```
Comment (jeongarmy) [inline apps/examples/testcase/le_tc/kernel/tc_libc_misc.c]: please fix output comments correctly. output is "SAMSUNG-TizenRT" only

File: apps/examples/testcase/le_tc/kernel/tc_libc_misc.c [hunk 14]
```diff
@@ -347,6 +349,72 @@ static void tc_libc_misc_vdbg(void)
...
+	free(buf);
+
+	TC_SUCCESS_RESULT();
+}
+
+/**
+ * @fn                  :tc_libc_misc_match
+ * @brief               :Simple shell-style filename pattern matcher
+ * @scenario            :Returns 1 (match) or 0 (no-match)
+ * @API's covered       :lib_match
```
Comment (jeongarmy) [inline apps/examples/testcase/le_tc/kernel/tc_libc_misc.c]: please modify it to match

File: apps/examples/testcase/le_tc/kernel/tc_libc_misc.c [hunk 15]
```diff
@@ -347,6 +349,72 @@ static void tc_libc_misc_vdbg(void)
...
+{
+	const char *msg = "tc_libc_misc_dumpbuffer";
+	unsigned char buffer[] = { 'S', 'A', 'M', 'S', 'U', 'N', 'G', '-', 'T', 'i', 'z', 'e', 'n', 'R', 'T' };
+	unsigned char *buf = NULL;
+	int idx;
+
+	/**
+	 * As lib_dumpbuffer returns void, there in no way to check the success or failure case
+	 * we need to manually check the dump output
+	 * Case-1: output: 53414d53554e472d54697a656e5254 "SAMSUNG-TizenRT"	only
```
Comment (jeongarmy) [inline apps/examples/testcase/le_tc/kernel/tc_libc_misc.c]: I guess you misunderstood what I said.
I meant  this comment should be "Case-1: output: SAMSUNG-TizenRT"
I didn't mean adding "only"

Comment (jeongarmy) [inline apps/examples/testcase/le_tc/kernel/tc_libc_misc.c]: I guess you understood what I said.
I didn't mean adding " " and only to comments so please remove these
